lib/onelogin/ruby-saml/response.rb

```diff
@@ -112,78 +151,326 @@ def status_message
+      # (returns the first node that matches the supplied xpath)
+      # @return [REXML::Element] Conditions Element if exists
+      #
       def conditions
         @conditions ||= xpath_first_from_signed_assertion('/a:Conditions')
       end
 
+      # Gets the NotBefore Condition Element value.
+      # @return [Time] The NotBefore value in Time format
+      #
       def not_before
         @not_before ||= parse_time(conditions, "NotBefore")
       end
 
+      # Gets the NotOnOrAfter Condition Element value.
+      # @return [Time] The NotOnOrAfter value in Time format
+      #
       def not_on_or_after
         @not_on_or_after ||= parse_time(conditions, "NotOnOrAfter")
       end
 
-      def issuer
-        @issuer ||= begin
-          node = REXML::XPath.first(document, "/p:Response/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
-          node ||= xpath_first_from_signed_assertion('/a:Issuer')
-          node.nil? ? nil : node.text
+      # Gets the Issuers (from Response and Assertion).
+      # (returns the first node that matches the supplied xpath from the Response and from the Assertion)
+      # @return [Array] Array with the Issuers (REXML::Element)
+      #
+      def issuers
+        @issuers ||= begin
+          issuers = []
+          nodes = REXML::XPath.match(document, "/p:Response/a:Issuer | /p:Response/a:Assertion/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
+          nodes.each do |node|
+            issuers << node.text if not node.text.nil?
+          end
+          issuers.uniq
+        end
+      end
+
+      # @return [String|nil] The InResponseTo attribute from the SAML Response.
+      #
+      def in_response_to
+        @in_response_to ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['InResponseTo']
         end
       end
 
+      # @return [String|nil] Destination attribute from the SAML Response
+      #
+      def destination
+        @destination ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['Destination']
+        end
+      end
+
+      # @return [Array] The Audience elements from the Contitions of the SAML Response.
+      #
+      def audiences
+        @audiences ||= begin
+          audiences = []
+          nodes = xpath_from_signed_assertion('/a:Conditions/a:AudienceRestriction/a:Audience')
+          nodes.each do |node|
+            if node && node.text
+              audiences << node.text
+            end
+          end
+          audiences
+        end
+     end
+
       private
 
-      def validate(soft = true)
-        valid_saml?(document, soft)      &&
+      # Gets the expected current_url
+      # (Right now we read this url from the Assertion Consumer Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
```

Comment:
This is either a copy-pasta comment, or this should really be addressed in this changeset.